### ISSUE TITLE
Sire::IO::AmberPrm topology file giving strange MD results

### ISSUE THREAD

**Issue by AdeleLip** [michellab/Sire#338]
I am running some simple production MD, and am seeing protein unfolding where there shouldn't be any. This is the code I run:
```python
import BioSimSpace as BSS
from shutil import copyfile

system = BSS.IO.readMolecules(['system_bash.prm7', 'system_bash.rst7'])
protocol = BSS.Protocol.Production(runtime=5*BSS.Units.Time.nanosecond, restart_interval=2500)
process = BSS.Process.Amber(system, protocol, exe='/home/adele/software/amber20_20.04/amber20/bin/pmemd.cuda')

process.start()
process.wait()
copyfile(f'{process.workDir()}/amber.nc', 'production_BSS.nc')
```
This results in my protein start to unfold (particularly near the N terminal, highlighted in blue):
![image](https://user-images.githubusercontent.com/64790565/106763825-2084c400-662f-11eb-82b4-04cca8674f6f.png)
I have also seen complete unfolding after 100 ns previously (which was what led me to investigate):
![Outlook-tvjmo5nz](https://user-images.githubusercontent.com/64790565/106764173-722d4e80-662f-11eb-889d-ecf5fd1ffc6c.png)

However, when I run the following:
```python
import BioSimSpace as BSS
from shutil import copyfile

system = BSS.IO.readMolecules(['system_bash.prm7', 'system_bash.rst7'])
protocol = BSS.Protocol.Production(runtime=5*BSS.Units.Time.nanosecond, restart_interval=2500)
process = BSS.Process.Amber(system, protocol, exe='/home/adele/software/amber20_20.04/amber20/bin/pmemd.cuda')

copyfile('system_bash.prm7', f'{process.workDir()}/amber.prm7')
copyfile('system_bash.rst7', f'{process.workDir()}/amber.rst7')

process.start()
process.wait()
copyfile(f'{process.workDir()}/amber.nc', 'production_BSS.nc')
```
I get a stable protein, as expected and as I've seen for MD runs outside of BSS. Is it possible that the reading/writing of the system is resulting in a topology different enough to destabilise protein on such a short timescale?

I'm including relevant files:
[file-diff.zip](https://github.com/michellab/BioSimSpace/files/5919298/file-diff.zip)

(I also tried this with only copying over the original topology file and using the BSS generated rst, which worked fine; this is pointing me towards the topology being an issue).

I am using the dev version of BSS downloaded yesterday (02/02), but the release version downloaded today (03/02) gave the same results. 

On import I get:
`/home/adele/anaconda3/envs/BSS-env/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: numpy.ndarray size changed, may indicate binary incompatibility. Expected 80 from C header, got 88 from PyObject
  return f(*args, **kwds)` 
but michellab/BioSimSpace#37 makes me think it shouldn't matter?

Might be unrelated (or a seprate issue), but when I run:
```python
protein = BSS.IO.readMolecules('open.pdb')
protein_parm_process = BSS.Parameters.ff14SB(protein.getMolecule(0))
protein_parm = protein_parm_process.getMolecule()
solvated = BSS.Solvent.tip3p(protein_parm.toSystem(), shell=10*BSS.Units.Length.angstrom, ion_conc=0.15, is_neutral=True)
minimisation_protocol = BSS.Protocol.Minimisation(steps=2000)
minimisation_process = BSS.Process.Amber(solvated, minimisation_protocol)
minimisation_process.start()
minimisation_process.wait()
minimised = minimisation_process.getSystem()
```
I get the following Sire issue:
`The property of type SireMol::AtomCoords is incompatible with the layout with UID {d01e83b8-690a-45b5-9e27-73ebb2651b85}`
`IncompatibleError: Unable to update 'coordinates' for molecule index '1'`
When I load the molecule fresh from the workDir I can continue. Additionally, putting a system parameterised/solvated in leap outside BSS through the same process does not raise an error. Let me know if you think this is related, or if this a separa/non issue.

**Comment by lohedges** [comment]
For reference, `ParmEd` preserves the original topology on write:

```python
import parmed as pd

ps =  parmed.load_file("system_bash.prm7", xyz="system_bash.rst7")
ps.save("system_parmed.prmtop")
```

_system_parmed.prmtop_:
```
%FLAG POINTERS
%FORMAT(10I8)
   35945      18   33545    2361    5225    3181   10531   10049       0       0
   67175   10793    2361    3181   10049      70     162     216      38       1
       0       0       0       0       0       0       0       1      24       0
       0
```

**Comment by lohedges** [comment]
I'm pretty sure that the issue is occurring on _read_, since the `parameters` properties for each molecule loaded from `system_bash.*` and `system_BSS.*` appear to be identical. A basic single-point energy comparison of the two systems (within Sire) also gives identical results.

**Comment by lohedges** [comment]
I have a small lead for the mismatched `NPHB` pointer. On read this is correctly flagged as 1 but on write it is 0.

```python
import BioSimSpace as BSS

# I've added a print statement into the parser to print the pointer.
s = BSS.IO.readMolecules(BSS.IO.glob("system_bash.*"))
NPHB = 1

BSS.IO.saveMolecules("test", s, "prm7")
NPHB = 0
```

Looking at the conditional [here](https://github.com/michellab/Sire/blob/devel/corelib/src/libs/SireIO/amberprm.cpp#L436), I'm not even sure if we support 10-12 hydrogen bond parameters? Perhaps this is the problem, i.e. we are not catching this properly so we are then making some incorrect assumptions for other records?

```cpp
       ...

       //amber stores the A and B coefficients as the product of all
        //possible combinations. We need to find the values from the
        // LJ_i * LJ_i values
        int idx = nb_parm_index[ ntypes * i + i  ];

        if (idx < 0)
        {
            //this is a 10-12 parameter
            throw SireError::unsupported( QObject::tr(
                    "Sire does not yet support Amber Parm files that "
                    "use 10-12 HBond parameters."), CODELOC );
        }
        ...
```

@chryswoods: Do you have any thoughts on this?

I'm not yet sure about the issue with the other pointers, but at least they so far seem to be consistent on read/write for the ones that I've tested.

**Comment by lohedges** [comment]
Yes, this looks like the issue. Here is the `NONBONDED_PARM_INDEX` from the `system_bash.prm7` file:

```
%FLAG NONBONDED_PARM_INDEX
%FORMAT(10I8)
       1       2       4       7      11      16      22      29      37      46
      56      67      79      92     106     121     137     154       2       3
       5       8      12      17      23      30      38      47      57      68
      80      93     107     122     138     155       4       5       6       9
      13      18      24      31      39      48      58      69      81      94
     108     123     139     156       7       8       9      10      14      19
      25      32      40      49      59      70      82      95     109     124
     140     157      11      12      13      14      15      20      26      33
      41      50      60      71      83      96     110     125     141     158
      16      17      18      19      20      21      27      34      42      51
      61      72      84      97     111     126     142     159      22      23
      24      25      26      27      28      35      43      52      62      73
      85      98     112     127     143     160      29      30      31      32
      33      34      35      36      44      53      63      74      86      99
     113     128     144     161      37      38      39      40      41      42
      43      44      45      54      64      75      87     100     114     129
     145     162      46      47      48      49      50      51      52      53
      54      55      65      76      88     101     115     130     146     163
      56      57      58      59      60      61      62      63      64      65
      66      77      89     102     116     131     147     164      67      68
      69      70      71      72      73      74      75      76      77      78
      90     103     117     132     148     165      79      80      81      82
      83      84      85      86      87      88      89      90      91     104
     118     133     149     166      92      93      94      95      96      97
      98      99     100     101     102     103     104     105     119     134
     150     167     106     107     108     109     110     111     112     113
     114     115     116     117     118     119     120     135     151     168
     121     122     123     124     125     126     127     128     129     130
     131     132     133     134     135     136     152     169     137     138
     139     140     141     142     143     144     145     146     147     148
     149     150     151     152     153      -1     154     155     156     157
     158     159     160     161     162     163     164     165     166     167
     168     169      -1     171
```

Note that the second to last entry is `-1`, hence the conditional above should have been triggered and an exception thrown. I'll try to figure out why this isn't the case.

**Comment by lohedges** [comment]
These are possible not being read because of the incorrect `NTYPES` pointer, i.e. 17 rather than 18. The loop is over `NTYPES` x `NTYPES` terms, i.e. 18 x 18 = 324. However, the code is actually looping over 17 x 17 = 289 records, so is missing both of the -1 entries, which are in the last 19 places.

Now to work out why `NTYPES` is incorrect.

**Comment by lohedges** [comment]
This probably explains why the other pointers / flags are incorrect too, since the are also built from looping over the incorrect number of terms. The `NTYPES` record is actually built from looping over the parameters that were read and working out total number of distinct LJ parameters, i.e.:

```cpp
        for (int i=0; i<params.count(); ++i)
        {
            const auto info = params.constData()[i].info();
            const auto ljs = params.constData()[i].ljs();

            QVector<qint64> mol_atom_types(info.nAtoms());

            for (int j=0; j<info.nAtoms(); ++j)
            {
                const LJParameter lj = ljs[ info.cgAtomIdx(AtomIdx(j)) ];

                int idx = ljparams.indexOf(lj);

                if (idx == -1)
                {
                    ljparams.append(lj);
                    mol_atom_types[j] = ljparams.count();
                }
                else
                {
                    mol_atom_types[j] = idx + 1;
                }
            }

            atom_types[i] = mol_atom_types;
        }

        // We now have all of the atom types - create the acoeff and bcoeff arrays.
        const int ntypes = ljparams.count();
```
I assume one term (possibly for those involving 10-12 hydrogen bonds) is being processed incorrectly, which is leading the the miscount.

For simplicity I guess we could just raise an exception whenever `NPHB` is non-zero. We should also probably add in checks that the number of records related to certain `FLAG`entries matches the expected `NTYPES` x `NTYPES`. If not, then we know that we haven't correctly identified the unique types in the system.

**Comment by chryswoods** [comment]
Yes, that sounds right. Sire does not support 10-12 terms as these (I thought) were an old relic of older force fields and no longer widely used. They would require a lot of work to support (a whole duplicate of 10-12 versions of the LJ terms and forcefields) and so I didn't think it was worth it.

I am surprised that the exception was not raised on read. It would be worth finding out why 10-12 terms are in this file, particularly as they would break some of our MD drivers (e.g. somd) and are not supported by our gromacs writer either (I think). 

@jmichel80 do you think 10-12 terms are needed or are making a comeback? Is it worth putting in the work to support them?

**Comment by lohedges** [comment]
Yes, I thought that they were only associated with some of the older force fields too. Thanks for the info regarding the lack of support for the other MD drivers.

I'll try to work out why the `NTYPES` pointer is being inferred incorrectly. (We read it, but then re-construct everything from the molecular topology generated from the other records.)

If we don't want to support 10-12 terms it would be easy enough to raise an exception whenever `NPHB` (`pointers[19]` in the code) is non-zero. This is read before any of the flags are processed so would be caught early.

It think it would also be good to add some self-consistency checks to the code too, since this would help us catch situations where we are misreading (or re-constructing) the data in the topology file.

**Comment by lohedges** [comment]
Actually, there are some self-consistency checks, e.g. here:

```cpp

    const int ntypes = pointers[1];  //number of distinct atom types

    if (ntypes <= 0)
        return;

    const int nphb = pointers[19];   //number of distinct 10-12 hydrogen bond pair types

    lj_data = QVector<LJParameter>(ntypes);
    auto lj_data_array = lj_data.data();

    auto acoeffs = float_data.value("LENNARD_JONES_ACOEF");
    auto bcoeffs = float_data.value("LENNARD_JONES_BCOEF");

    auto hbond_acoeffs = float_data.value("HBOND_ACOEF");
    auto hbond_bcoeffs = float_data.value("HBOND_BCOEF");

    auto nb_parm_index = int_data.value("NONBONDED_PARM_INDEX");

    qDebug() << "NONBONDED_PARM_INDEX:" << nb_parm_index.count();
    qDebug() << "NTYPES:" << ntypes;

    if (acoeffs.count() != bcoeffs.count() or
        acoeffs.count() != (ntypes*(ntypes+1))/2)
    {
        throw SireIO::parse_error( QObject::tr(
                "Incorrect number of LJ coefficients for the number of specified "
                "atom types! Should be %1 for %2 types, but actually have "
                "%3 LJ A-coefficients, and %4 LJ B-coefficients")
                    .arg((ntypes*(ntypes+1))/2)
                    .arg(ntypes)
                    .arg(acoeffs.count())
                    .arg(bcoeffs.count()), CODELOC );
    }

    if (nb_parm_index.count() != ntypes*ntypes)
    {
        throw SireIO::parse_error( QObject::tr(
                "Incorrect number of non-bonded parameter indicies. There should "
                "be %1 indicies for %2 types, but actually have %3.")
                    .arg(ntypes*ntypes)
                    .arg(ntypes)
                    .arg(nb_parm_index.count()), CODELOC );
    }

    if (hbond_acoeffs.count() != nphb or
        hbond_bcoeffs.count() != nphb)
    {
        throw SireIO::parse_error( QObject::tr(
                "Incorrect number of HBond parameters. There should be "
                "%1 such parameters, but the number of HBond A coefficients is "
                "%2, and the number of B coefficients is %3.")
                    .arg(nphb)
                    .arg(hbond_acoeffs.count())
                    .arg(hbond_bcoeffs.count()), CODELOC );
    }
```

On read, this prints:

```
NONBONDED_PARM_INDEX: 324
NTYPES: 18
```

This means that it _is_ looping over the correct number of entries, i.e. it is using the pointers value rather than that inferred from the unique LJ parameters. I'll figure out why it isn't seeing the -1 entries (presumably the loop is wrong) since this would trigger the exception.

**Comment by lohedges** [comment]
Looking again at the code the loop is only over diagonal elements of the matrix, so we're missing the -1 terms. (As such, I'm not sure if we'd ever detect the presence of 10-12 terms.) [This](https://parmed.github.io/ParmEd/html/_modules/parmed/amber/_amberparm.html) code from ParmEd has routines for looping over all interactions, detecting non-zero 10-12 terms, etc.

If we don't care about the off-diagonal terms then we could just add an additional loop inside [build_lj](https://github.com/michellab/Sire/blob/devel/corelib/src/libs/SireIO/amberprm.cpp#429) that loops over them and raises the exception if any index is -1.

**Comment by lohedges** [comment]
Adding something like the following as a check:

```cpp
    // The build_lj function above only considers diagonal elements of the
    // NONBONDED_PARM_INDEX matrix. Here we loop over the off-diagonal elements
    // to check for 10-12 parameters, which are currently unsupported.

    // The matrix is symmetric, so perform a triangular loop over off-diagonal
    // elements.
    for (int i=0; i<ntypes; ++i)
    {
        for (int j=i+1; j<ntypes; ++j)
        {
            int idx = nb_parm_index[ ntypes * i + j  ];

            if (idx < 0)
            {
                auto a = hbond_acoeffs[idx];
                auto b = hbond_bcoeffs[idx];

                if ((a > 1e-6) and (b > 1e-6))
                {
                    //this is a 10-12 parameter
                    throw SireError::unsupported( QObject::tr(
                            "Sire does not yet support Amber Parm files that "
                            "use 10-12 HBond parameters."), CODELOC );
                }
            }
        }
    }
```

However, looking at the `system_bash.prm7` file, the values of the coefficients are:

```
%FLAG HBOND_ACOEF
%FORMAT(5E16.8)
  0.00000000E+00
%FLAG HBOND_BCOEF
%FORMAT(5E16.8)
  0.00000000E+00
```

Since they are zero, there aren't any non-zero 10-12 terms for this system, so we should be okay. ([SireUnitTests](https://github.com/michellab/SireUnitTests) contains several top files with -1 entries in the `NONBONDED_PARM_INDEX` flags, but also with zero terms for the coefficients.)

Will try to work out why the other pointers are incorrect.

**Comment by lohedges** [comment]
Hmm, I've realised the writing out with ParmEd was simply saving the original file to back to disk. Going via an OpenMM system loses the `NPHB` pointer for this system, so it's clearly redundant. (For this system, at least.) Sorry about that.

```python
import parmed as pd

ps0 =  parmed.load_file("system_bash.prm7", xyz="system_bash.rst7")
omm_system = ps0.createSystem()
ps1 = pd.openmm.load_topology(ps0.topology, system)
ps1.save("system_parmed.prmtop")
```

Here are the pointers for the original system and the BSS and ParmEd conversions:

_system_bash.prm7_
```sh
%FLAG POINTERS
%FORMAT(10I8)
   35945      18   33545    2361    5225    3181   10531   10049       0       0
   67175   10793    2361    3181   10049      70     162     216      38       1
       0       0       0       0       0       0       0       1      24       0
       0
```

_system_BSS.prm7_
```sh
%FORMAT(10I8)
   35945      17   33545    2361    5225    3181    9906    8439       0       0
   67175   10793    2361    3181    8439      37      44     228      38       0
       0       0       0       0       0       0       0       1      24       0
       0       0       0
```

_system_parmed.prmtop_:
```
%FLAG POINTERS
%FORMAT(10I8)
   35945      17   33545    2361    5225    3181   10531   10049       0       0
   67175   10793    2361    3181   10049      37      44   12269       1       0
       0       0       0       0       0       0       0       0      24       0
       0
```

Note that the reconstructed ParmEd topology disagrees in the `NATYP` pointer (it is 1, when it should be 38). It also claims that there are 12269 unique dihedral types, rather than 216. However, it now agrees with Sire that there are 17 rather than 18 unique types, (`NTYPES`), 37 rather than 70 unique bonds (`NUMBND`) and 44 rather than 162 unique angles (`NUMANG`).

It would be interesting to know whether the ParmEd file blows up in simulation too, which might point to their being something funky with the original system. (At least in the sense of it being a hard system to parse correctly.) The fact that, apart from the dihedrals,  it agrees more closely with Sire after conversion makes me think that something is wrong, or that there is redundant information in the original topology.

I'll keep digging on the other terms.

**Comment by lohedges** [comment]
On read the dihedral terms are correct, in the sense that counting the total number of dihedrals plus impropers with and without hydrogen gives the correct values of 10531 and 10049 respectively, i.e. 10027 + 504 and 9620 + 429. However, the values that are produced on write, 9906 and 8439, aren't simply the value of _just_ the dihedral terms with and without hydrogen, i.e. missing the impropers, which would be 10027 and 9620.

**Comment by lohedges** [comment]
Something that I should have tried earlier is a single-point energy comparison between the the three topology files using `sander`, rather than just doing this internally within Sire. Here I'm performing a default, single-step, minimisation using BioSimSpace and looking at the energy log file `amber.nrg`. I've just replaced `amber.prm7` in the working directory with the appropriate topology file, and re-run the minimisation. (Importantly this is copied across by hand, not written by BioSimSpace.)

Here are the results:

```
sander -O -i amber.cfg -p amber.prm7 -c amber.rst7 -o stdout -r amber.crd -inf amber.nrg
```


_system_bash.prm7_:
```
   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1      -1.1623E+05     1.4600E+01     1.0911E+02     C        3537

 BOND    =      867.4181  ANGLE   =     2384.8699  DIHED      =     3746.3088
 VDWAALS =    12410.6216  EEL     =  -148008.4340  HBOND      =        0.0000
 1-4 VDW =     1063.6145  1-4 EEL =    11302.3846  RESTRAINT  =        0.0000
```

_system_BSS.prm7_:
```
   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1      -1.1623E+05     1.4600E+01     1.0911E+02     C        3537

 BOND    =      867.4180  ANGLE   =     2384.8698  DIHED      =     3746.3089
 VDWAALS =    12410.6217  EEL     =  -148008.4341  HBOND      =        0.0000
 1-4 VDW =     1063.6145  1-4 EEL =    11302.3846  RESTRAINT  =        0.0000
```

_system_parmed.prm7_:
```
   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1      -1.1623E+05     1.4600E+01     1.0911E+02     C        3537

 BOND    =      867.4180  ANGLE   =     2384.8698  DIHED      =     3746.3089
 VDWAALS =    12410.6217  EEL     =  -148008.4341  HBOND      =        0.0000
 1-4 VDW =     1063.6145  1-4 EEL =    11302.3846  RESTRAINT  =        0.0000
```

As you can see, despite the differences in the topology files, the energies are in near perfect agreement. BioSimSpace and ParmEd agree _exactly_, whereas the original `system_bash.prm7` file differs only in the fourth decimal place for terms in the first two lines.

Differences in topology file records are hard to interpret without looking at single-point energy tests such as those above, since Sire performs a lot of clever de-duplication of the records, such as for dihedral terms. This would explain why the pointers count for those records is lower, yet the energy is the same. (There can be a lot of redundancy in a topology file.)

Given the above, I'm at a loss to explain why you are seeing such markedly different results using the BioSimSpace topology. Is the behaviour repeatable across multiple runs? What happens if you use the ParmEd topology file? If only the BioSimSpace topology reproducibly gives weird results, then the only thing that I can think of is that there are other records in the topology file that are being lost on write that are causing incompatibility with some of the default options that we use for our other protocols, e.g.  `BioSimSpace.Protocol.Production`.

**Comment by lohedges** [comment]
I just realised that I had copied across the binary `system_bash.rst7` file to use as the coordinates file `amber.rst7`. If you use the `Sire.IO.AmberPrm7` text file written by BioSimSpace (which you said hasn't caused issues), then the energies computed using all three topology are _identical_, i.e. the differences in the fourth decimal places disappear.

**Comment by AdeleLip** [comment]
I have tried a lot of different bash+copy/BSS combinations to arrive at the conclusion it's something to do with topology, so this issue has happened for multiple runs.

I also used BSS for production runs for older systems I had setup with ff19SB. I know now BSS doesn't support ff19SB, but loading the topology into AMBER's cpptraj and saving it again seemed to remove the CMAP parameters and the system could then be loaded into BSS (which isn't entirely right but it worked for the time being). Those runs were alright for the 100-200 ns timescale using all the files created by BSS in terms of protein stability. However, the system did show some behaviour that was different (mainly some strange loop dynamics). This doesn't happen now that I copy over files.

All this makes me wonder if there is just something with the system itself that is just really sensitive to some small change, since I assume noone else has complained about protein unfolding so far...

I can do some more tests with ParmEd once I one of my GPUs frees up tomorrow.

**Comment by jmichel80** [comment]
If the CMAP terms were omitted surely the dihedrals would be completely wrong, leading to erratic behavior, especially on a longer MD timescale ?

**Comment by lohedges** [comment]
Thanks for the update. I'd certainly be interested in the ParmEd comparison if you get a chance.

Just to check:

> I also used BSS for production runs for older systems I had setup with ff19SB. I know now BSS doesn't support ff19SB, but loading the topology into AMBER's cpptraj and saving it again seemed to remove the CMAP parameters and the system could then be loaded into BSS (which isn't entirely right but it worked for the time being). Those runs were alright for the 100-200 ns timescale using all the files created by BSS in terms of protein stability. However, the system did show some behaviour that was different (mainly some strange loop dynamics). This doesn't happen now that I copy over files.

Is the comparison here between the original ff19SB files with CMAP records and those generated by BioSimSpace after the loading and saving the CCPTRAJ files that stripped those records, or between the CCPTRAJ files themselves and those written by BioSimSpace? If it was the former, then I imagine that the CMAP terms in the original files are stabilising backbone fluctuations that you are seeing with BioSimSpace. If it's the latter, then it suggests that something is going wrong with the BioSimSpace generated topology. (It might be good to check single-point energies for these files too.)

**Comment by AdeleLip** [comment]
yes, I do think remoivng the CMAP parameters must've messed with the system somehow. However that happened everytime I'd load the ff19SB system into cpptraj for processing, it would just say those are ignored, loading then immediately saving the topology without doing anything else to it removed them, and the simulations ran (kind of?) okay so it didn't alarm me as much as it should have.

The comparison is between (1) original ff19SB ran by hand, (2) ff19SB with CMAP automatically removed by cpptraj and ran in BSS and now also (3) ff14SB ran in BSS but while copying files. The behaviour I see in (1) and (3) is the same, while (2) is different but as Julien said it might be due to the whole system being a mess without CMAP. I can go back to those systems with ff19SB and run a short test simulation by hand to see what happens with cpptraj generated topologies without CMAP ran by hand.

**Comment by lohedges** [comment]
Okay, that explains the difference then. Only the `system_bash.prm7` file doesn't include any CMAP records?

Looking at the CCPTRAJ [GitHub page](https://github.com/Amber-MD/cpptraj/commit/e5a329a3adae44703a0170b08d35245bfa5e6e3c) it looks like support for CMAP records was only added in September of last year, so it might not be in the most recent AmberTools. I know that ParmEd supports the records, although I'm not sure if this is just read/write to the same format, or on conversion too.

**Comment by AdeleLip** [comment]
Yes, so the process of getting an ff19SB system into BSS was:
1. load topology into cpptraj
2. save the topology (no changes made)
3. load the new topology and unchanged `.rst7` files
4. proceed with production

The `system_bash.prm7` file (and all files I've uploaded here) is different, however. It has been prepared with ff14SB to stay consistent with what BSS supports, since I made all those files while trying to identify what was going wrong.

**Comment by lohedges** [comment]
Thanks for the clarification. As @jmichel80 says, I would certainly expect the systems without CMAP terms (for a forcefield that uses them) to be unstable on long time scales if those terms were removed. As for the ff14SB non-CMAP system, given that the single-point energies agree, I can't think what could have gone wrong.

I had another look at the files and noticed that there is a `BOX_DIMENSIONS` record in the `prm7` files. This differs between `system_bash.prm7` and `system_bss.prm7`:

_system_bash.prm7_:
```
%FLAG BOX_DIMENSIONS
%FORMAT(5E16.8)
  9.00000000E+01  9.22133330E+01  7.39995300E+01  6.71360020E+01
```

_system_BSS.prm7_:
```
%FLAG BOX_DIMENSIONS
%FORMAT(5E16.8)
  9.00000000E+01  8.49602111E+01  6.81790310E+01  6.18553599E+01
```

However, there must also be box information in the `*.rst7 `files, since when I load them the `space` property of the systems are identical:

```python
import BioSimSpace as BSS

s_bash = BSS.IO.readMolecules(BSS.IO.glob("system_bash.*"))
s_bash._sire_object.property("space")
PeriodicBox( ( 84.9602, 68.179, 61.8554 ) )

s_bss = BSS.IO.readMolecules(BSS.IO.glob("system_BSS.*"))
s_bss._sire_object.property("space")
PeriodicBox( ( 84.9602, 68.179, 61.8554 ) )
```

I think that this record in the topology file is redundant, and that the information in the RST7 file should take precedence. (This is the case with the Sire parsers when constructing a system.) However, perhaps this isn't the case when running AMBER? It might be worth trying a simulation where you copy the dimensions from the `system_bash.prm7` file into the `system_BSS.prm7` file. That said, the `IFBOX` pointer is 0 so the record _should_ be ignored according to the [AMBER docs](https://ambermd.org/FileFormats.php).

(I note that ParmEd strips the BOX_DIMENSIONS flag from the topology file on write and the values that it writes to a RST7 file are consistent with Sire.)

**Comment by AdeleLip** [comment]
I ran a simulaiton test with the box dimensions copied over from `system_BSS.prm7` as follows:
```python
import BioSimSpace as BSS
from shutil import copyfile
system = BSS.IO.readMolecules(['system_bash.prm7', 'system_bash.rst7'])
protocol = BSS.Protocol.Production(runtime=5*BSS.Units.Time.nanosecond)
process = BSS.Process.Amber(system, protocol, exe='/home/adele/software/amber20/bin/pmemd.cuda')
copyfile('system_bash.prm7', f'{process.workDir()}/amber.prm7')
#Replace %FLAG BOX_DIMENSIONS with the one from system_BSS.prm7
process.start()
```
the system was stable after 5 ns.

I have also tried running the same simulation using ParmEd topology via OpenMM as you've done before:
```python
import BioSimSpace as BSS
import parmed as pmd
from shutil import copyfile
top0 = pmd.load_file('system_bash.prm7', xyz='.system_bash.rst7')
parmed_system = top0.createSystem()
top1 = pmd.openmm.load_topology(top0.topology, parmed_system)
top1.save('system_parmed.prm7', format='amber')
system = BSS.IO.readMolecules(['system_bash.prm7', 'system_bash.rst7'])
protocol = BSS.Protocol.Production(runtime=5*BSS.Units.Time.nanosecond)
process = BSS.Process.Amber(system, protocol, exe='/home/adele/software/amber20/bin/pmemd.cuda')
copyfile('system_parmed.prm7', f'{process.workDir()}/amber.prm7')
process.start()
```
however, I get the following error:
`| ERROR:   the combination ntb != 0, ntp != 0, ifbox == 0 is not supported!
 Input errors occurred. Terminating execution.`
In section `1.  RESOURCE   USE` of the output file it gives:
```
 getting new box info from bottom of inpcrd
 NATOM  =   35945 NTYPES =      17 NBONH =   33545 MBONA  =    2361
 NTHETH =    5225 MTHETA =    3181 NPHIH =   10531 MPHIA  =   10049
 NHPARM =       0 NPARM  =       0 NNB   =   67175 NRES   =   10793
 NBONA  =    2361 NTHETA =    3181 NPHIA =   10049 NUMBND =      37
 NUMANG =      44 NPTRA  =   12269 NATYP =       1 NPHB   =       0
 IFBOX  =       0 NMXRS  =      24 IFCAP =       0 NEXTRA =       0
 NCOPY  =       0
```
however, the inpcrd file was not changed. And if I run the above process omitting the `copyfile` line, the output gives `IFBOX = 1`. I'm including the topology, coordinate and output files [here](https://github.com/michellab/Sire/files/6004762/parmed-test.zip). The coordinate files are exactly the same.

The ParmEd I used was the default that came with my Amber20 installation, could that be an issue? If you managed to run the energy tests with `sander` using the topology you produced with ParmEd and could upload the topology here, I can try to repeat it with that file.

**Comment by lohedges** [comment]
Thanks for the info. Good to know that changing the box dimensions doesn't have an effect. (It shouldn't.) Here is the ParmEd topology that was generated via conversion to / from an OpenMM system:

[system_parmed.prm7.txt](https://github.com/michellab/Sire/files/6008869/system_parmed.prm7.txt)

**Comment by AdeleLip** [comment]
I've compare the two ParmEd system files and they are identical. I have also finished the 5 ns test with the working ParmEd topology (`system_parmed_working.prm7`), and the system is stable.
[Here](https://github.com/michellab/Sire/files/6009330/parmed-test-updated.zip) are the files. I've also included the final frames of a simulation with using default BSS topology (`system_BSS_final.pdb`) and the modified ParmEd topology (`system_parmed_final.pdb`) for reference.

**Comment by AdeleLip** [comment]
I have pulled 10 frames from a steering trajectory as discussed [here](https://github.com/michellab/Sire/files/6337301/ptp1b-frames.zip). The folder also includes the system topology.

This trajectory is quite long (150 ns) and I was seeing unfolding within a few ns in my previous tests. Let me know if you want frames within the first 5-10 ns and I can get those for you as well.

**Comment by lohedges** [comment]
Just to check: The `system.prm7` file included in the directory looks different to the `system_bash.prm7` from the post, e.g. different number of atoms, etc.

`system.prm7`
```
%FLAG POINTERS
%FORMAT(10I8)
   40270      21   37769    2458    5397    3312   10879   10405       0       0
   73621   12193    2458    3312   10405      73     170     217      41       1
       0       0       0       0       0       0       0       1      24       0
```

`system_bash.prm7`
```
%FLAG POINTERS
%FORMAT(10I8)
   35945      18   33545    2361    5225    3181   10531   10049       0       0
   67175   10793    2361    3181   10049      70     162     216      38       1
       0       0       0       0       0       0       0       1      24       0
       0
```

Is this just the same system with, e.g., a different number of water molecules? Also, is this the original topology, i.e. from you setup pipleline, or the one written by BioSimSpace. Since this differs from the `system_bash.prm7` I'll need the equivalent file to be able to make comparisons, i.e. the one from `tLEaP`, not BioSimSpace.

**Comment by AdeleLip** [comment]
Sorry, I forgot which system I had provided originally. This is PTP1B +  a peptide substrate. I can get snapshots from just PTP1B, as well as snapshots from an unstable trajectory.

For reference this topology is written by leap.

**Comment by lohedges** [comment]
That's okay. As long as it's the `tLEaP` files and the system is unstable, that's all we need.

**Comment by lohedges** [comment]
Single point energies agree along the entire trajectory.  [This](https://github.com/michellab/Sire/files/6351017/single_point_check.tar.gz) archive contains some scripts to perform the checks (data already included). To run the single-point calculations, update the path to `sander` in `single_point.sh` and run it with something like:

```
./single_point.sh leap
```

Once energy files are generated we can compare them using `diff3` (assuming it is installed). The `check_energies.sh` script will do this for all frames and raise an error if the energies disagree and tell you what frame mismatched.

**Comment by AdeleLip** [comment]
[Here](https://github.com/michellab/Sire/files/6351462/unstable-system-frames.zip) are 10 frames from an unstable 5 ns snapshot. I loaded the same system as from yesterday and ran a default 5 ns Amber production process. I also included the topology BSS produced just in case. Maybe this will help to understand why this is happening...

I will also run short setups and production from other PDBs and will let you know how that goes.

**Comment by lohedges** [comment]
Sorry, realised I forgot to upload the [amber.cfg](https://github.com/michellab/Sire/files/6351516/amber.cfg.txt) file, which should be placed in the `min` directory. Energies also agree along the unstable trajectory.

**Comment by jmichel80** [comment]
hi @AdeleLip might be a studid question, the frames are not imaged ? are you sure the simulation is done in a periodic box ? It does look like a water droplet evaporating, leaving a protein in vacuum to unfold.

**Comment by AdeleLip** [comment]
I have been using the config files generated by BSS, so I assumed periodic conditions were applied. The file contents are:
```
Production.
 &cntrl
  ig=-1,
  ntx=1,
  ntxo=1,
  ntpr=2500,
  ntwr=2500,
  ntwx=2500,
  irest=0,
  dt=0.002,
  nstlim=2500000,
  ntc=2,
  ntf=2,
  ntt=3,
  gamma_ln=2,
  cut=8.0,
  tempi=300.00,
  temp0=300.00,
  ntp=1,
  pres0=1.01325,
 /
```
from the Amber20 manual:
> ntb    This variable controls whether or not periodic boundaries are imposed on the system
> ...
> = 2 constant pressure (default when ntp > 0)
> If NTB is nonzero then there must be a periodic boundary in the topology file.

I'm not 100% sure where in the topology file periodic boundaries are indicated, but if it's `DIHEDRAL_PERIODICITY` then I'm seeing a lot of differences between the original and the BSS topology. Could that be it?

**Comment by lohedges** [comment]
The topology file has a `%FLAG BOX_DIMENSIONS` record, although this is now deprecated. Instead, box dimensions are read from the coordinate file.

I'll check the AMBER configs. This was one of the first things I set up, so I might have missed some options, or they might not be set correctly when other options are turned on / off. I know that I set `ntb=0` if there is no box.

**Comment by lohedges** [comment]
It looks like I don't explicitly set `ntb` if constant pressure is turned on, since it's set to `2` by default if `ntp>0`. (As in the docs that you posted.) However, I don't seem to set if it it's not constant pressure. I think it's `ntb=1` by default, but will check in the manual.

**Comment by jmichel80** [comment]
Might be a red herring, you need iwrap=1 (default 0) to wrap coordinates in the minimum image, see p 342 of https://ambermd.org/doc12/Amber20.pdf

@AdeleLip what happens if you run an MD simulation on the BSS prm7/rst7 inputs using a default config file from a relevant Amber20 tutorial ?

**Comment by lohedges** [comment]
The difference in the `DIHEDRAL_PERIODICITY` could be due to the ordering of the dihedral terms. I would have thought that that would surely lead to a difference in the single point energies, though.

**Comment by lohedges** [comment]
I built the configs using some _best practice_ from a few tutorials, but it would be good to test against some sane defaults from elsewhere to see if anything inconsistent.

**Comment by lohedges** [comment]
Is it preferable to set `iwrap=1`? I would have thought that it would be possible in post-processing and some people might want actual coordinates to compute displacements, etc. (If image numbers aren't available.)

**Comment by AdeleLip** [comment]
I just ran a test with BSS topology and the following cfg from [this tutorial](https://ambermd.org/tutorials/basic/tutorial14/index.php):
```
Production.
 &cntrl
   imin=0, irest=1, ntx=5, 
   ntpr=2500, ntwx=2500, ntwr=2500, nstlim=2500000, 
   dt=0.002, ntt=3, tempi=300, 
   temp0=300, gamma_ln=1.0, ig=-1, 
   ntp=1, ntc=2, ntf=2, cut=9, 
   ntb=2, iwrap=1, ioutfm=1,
 /
```
after 1.5 ns I see unfolding.

**Comment by lohedges** [comment]
I guess you're using `pmemd` here. Could you run the single-point calculations using that? I wonder if there are some rounding issues that are only apparent on the GPU?

**Comment by AdeleLip** [comment]
You're on to something here. I get small differences in energies for all frames (
[pmemd_energies.zip](https://github.com/michellab/Sire/files/6352387/pmemd_energies.zip)). I used pmem.cuda. From a quick inspection it looks like the differences are in 1-4 VDW and EEL energies.

**Comment by jmichel80** [comment]
It could be useful to run an OpenMM MD simulation on the same inputs to check numerical stability.

**Comment by AdeleLip** [comment]
Do you mean single point energies or full simulations? I was using this same system for openMM sMD testing and had no issues using Amber topologies.

**Comment by jmichel80** [comment]
ah yes that's a huge difference in 1,4 EEL energies

Something wrong with the 1,4 scaling factors ?

**Comment by lohedges** [comment]
Thanks for the update, that's really helpful. I feel like we're getting somewhere.

I see that `ParmEd` agrees with the `tLEaP` files, so it's only the ones generated by `Sire.IO.AmberPrm` that are giving weird results. I also note that the `ParmEd` and `tLEaP` topology files give the same energies (within error) using the `frame_0.rst7` coordinates with both `sander` and `pmemd.cuda`, i.e.

```
   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1      -1.3060E+05     1.4601E+01     1.2988E+02     N        3128

 BOND    =      935.0701  ANGLE   =     2453.6861  DIHED      =     3859.7638
 VDWAALS =    14410.2600  EEL     =  -165042.1226  HBOND      =        0.0000
 1-4 VDW =     1087.9674  1-4 EEL =    11691.6519  RESTRAINT  =        0.0000
```

In contrast, for `Sire.IO.AmberPrm` we mach with `sander` but have the following with `pmemd.cuda`:

```
   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1      -1.4274E+05     1.4618E+01     1.3410E+02     N        3128

 BOND    =      935.0701  ANGLE   =     2453.6861  DIHED      =     3859.7643
 VDWAALS =    14410.2669  EEL     =  -165042.1627  HBOND      =        0.0000
 1-4 VDW =      941.7919  1-4 EEL =     -299.4411  RESTRAINT  =        0.0000
```

(Note that the `pmemd.cuda` files provided referred to frames from the _stable_ trajectory.)

The fact that the `Sire.IO.AmberPrm` topology gives different energies at step zero implies that this isn't an error that is accumulated as the simulation progresses, rather an error that is present from the outset.

Pardon my ignorance, but are there terms in the topology file that are only used by `pmemd`, or is it just that the potential calculation is performed in a different. way. I'm just trying to pinpoint what parts of the topology file to focus on. From the energies you would assume that there is somethng wrong with the 1-4 factors (as @jmichel80 suggests), but the energies for these terms are the same when using `sander`.

It would be good to know if this issue is unique to this system, or a more general problem.

**Comment by lohedges** [comment]
I guess we need to check `SCEE_SCALE_FACTOR` and `SCNB_SCALE_FACTOR` terms. Since these apply to dihedrals, I wonder if there is an issue with regards to how these potential terms are accumulated on the GPU. I note this in the Sire code:

```
//get the NB14 scaling factor for this dihedral. This should be set
        //for only the first dihedral found that has this pair, otherwise
        //we would risk Amber double-counting this parameter
```

Perhaps this double counting will never occur with `sander`, but could be problematic with `pmemd`.

I think we also treat represent the dihedral terms slightly differently (although equivalently, mathematically) and perhaps this causes issues with the way the potential is computed on the GPU?

**Comment by jmichel80** [comment]
Perhaps that would explain the differences on the 3rd or 4th decimal for DIHED, EEL, VDWAALS but I doubt the 1,4 terms, the difference is very large. a parsing error in pmemd is more likely. Check if the same behaviour occurs with pmemd.mpi 

Comparing the SCEE_SCALE_FACTOR and SCNB_SCALE_FACTOR entries I see that system_BSS.prm7 has the first 19 entries set to 0.0 whereas system_leap.prm7 and system_parmed.prm7 only have the last 4 entries set to 0.0 .  

Related to issues we encountered a few years ago when debugging the Sire prm7 writer ?

https://github.com/michellab/BioSimSpace/issues/47
https://github.com/michellab/Sire/issues/233

Perhaps we have to dig in in the sander and pmemd sources to understand why they would handle differently the same input prm7.
Note this comment in pmemd source code prmtop_dat.F90 
    !We found the SCEE scale factor data so read it in, there should be
    !a value for each dihedral type. Note while there is one for each
    !dihedral type not all are actually used in the calculation since
    !1-4's are only done over unique dihedral terms. If multiple dihedrals
    !for a given set of atom types exist, with different pn's for example
    !then the last value of SCEE/SCNB should be used.

Perhaps pmemd makes implicit assumptions that sander doesn’t.

It would make sense that the issue is with multi term dihedrals as these are more frequently used for backbone torsions. Messing up the backbone torsions would explain the unfolding behaviour reported by Adele.

**Comment by lohedges** [comment]
Thanks for the insight, much appreciated. The ordering and number of terms will be different due to the way that we treat dihedrals. Since the parser originally worked in parallel we choose to sort dihedral terms before writing to ensure consistent ordering. (This allows us to check files using diffs, and also preserve the topology on read/write round trips.) I believe that there will also be a different number of entries due to the way that we handle multi-part dihedrals and impropers.

For the parallel part: I think we've already disabled parallel parsing due to a subtle issue discovered by Paolo that was never resolved. As a quick check, I can just turn off the sorting for dihedrals to see if that changes things. (If we aren't parsing in parallel then the ordering will be consistent anyway.)

Looking at the other thread you linked to (and the thread linked from that) it does seem that the issue could be related. [Here](https://github.com/michellab/Sire/issues/233) is the correct issue discussion in the Sire repository. It seems that we have some self-consistency problems with our issue tracking and closing. (Why I prefer to keep things in one thread in the repo to which it corresponds :-) ) As such, I'm not sure if that issue was ever suitably resolved. I'll go back through the discussion to see if there is any info that might give us more of an insight into this problem.

I might try to recruit @chryswoods tomorrow morning, if he has time.

Cheers.

**Comment by lohedges** [comment]
Just a note from the other issue thread regarding the difference in the number of terms in the POINTERS record:

> Be careful chasing "POINTERS" - there will be differences because we obtain a different number of 1-4 parameters, dihedral parts and CLJ parameters when reading from a system. This is because there are different equivalent ways of working this out, and so we cannot preserve the exact information during a AmberPrm => System => AmberPrm conversion.

**Comment by lohedges** [comment]
I've just played around with the sorting and this looks like it is required, since I don't get consistent energies with `sander` if I disable it.

**Comment by jmichel80** [comment]
I suggest to also post a query on the amber mailing list, their developers may want to know why a prm7 input gives inconsistent energies between sander and pmemd. Check also pmemd on CPU as the issue may be with the cuda kernel only

**Comment by lohedges** [comment]
I've just checked regular CPU `pmemd` and the issue still persists. The `Sire.IO.AmberPrm` topology gives the same energies as with `pmemd.cuda`, so it's not related to the GPU.

**Comment by lohedges** [comment]
Looking above at the comment from the `pmemd` source, this seems to constrast with the comment in Sire:

_pmemd_
```
!We found the SCEE scale factor data so read it in, there should be
!a value for each dihedral type. Note while there is one for each
!dihedral type not all are actually used in the calculation since
!1-4's are only done over unique dihedral terms. If multiple dihedrals
!for a given set of atom types exist, with different pn's for example
!then the last value of SCEE/SCNB should be used.
```

_Sire_
```
//get the NB14 scaling factor for this dihedral. This should be set
        //for only the first dihedral found that has this pair, otherwise
        //we would risk Amber double-counting this parameter
```

Sire is using the first dihedral as its reference, `pmemd` is using the last. I'll try changing the Sire code (perhaps writing the term for all pairs as a start) and post an updated topology.

**Comment by lohedges** [comment]
Actually, I think Sire takes care to ensure that only _unique_ dihedral terms are written to the file, so this shouldn't be an issue.

**Comment by lohedges** [comment]
I've had a look through the `ParmEd` code and they seem to have a fix for `pmemd`. In particular:

```python
# pmemd likes to skip torsions with periodicities of 0, which may be
        # present as a way to hack entries into the 1-4 pairlist. See
        # https://github.com/ParmEd/ParmEd/pull/145 for discussion. The solution
        # here is to simply set that periodicity to 1.
        inst._cleanup_dihedrals_with_periodicity_zero()
...

def _cleanup_dihedrals_with_periodicity_zero(self):
        """
        For torsions with only a single term and a periodicity set to 0, make sure pmemd still
        properly recognizes the necessary exception parameters. update_dihedral_exclusions will
        make sure that if a dihedral has a type pn0 *and* ignore_end is set to False (which means
        that it is required to specify exclusions), then it is the *only* torsion between those
        atoms in the system. This allows us to scan through our dihedrals, look for significant
        terms that have pn==0, and simply add another dihedral with pn=1 and k=0 to ensure that
        pmemd will always get that exception correct
        """
        new_dihedrals = []
        for dih in self.dihedrals:
            if dih.ignore_end or dih.type.per != 0:
                continue
            # If we got here, ignore_end must be False and out periodicity must be 0. So add
            # another dihedral
            dt = DihedralType(0, 1, 0, dih.type.scee, dih.type.scnb, list=self.dihedral_types)
            self.dihedral_types.append(dt)
            new_dihedrals.append(
                Dihedral(dih.atom1, dih.atom2, dih.atom3, dih.atom4, improper=dih.improper,
                         ignore_end=False, type=dt)
            )
            # Now that we added the above dihedral, we can start ignoring the end-group interactions
            # on this dihedral
            dih.ignore_end = True
        if new_dihedrals:
            self.dihedrals.extend(new_dihedrals)
```

Perhaps this is our issue?

**Comment by lohedges** [comment]
We don't have any zero periodicity single-term dihedrals, so this isn't the problem.

**Comment by lohedges** [comment]
I've tried a different protein-ligand system (one of Dom's funnel examples) and the energies also differ for the same terms using `pmemd`, so this appears to be a general issue. In the workshop we will use these files with `OpenMM`, which parses things with `ParmEd`. I'm not sure we'll there will be potential issues there too.

**Comment by jmichel80** [comment]
we can establish this by computing single point energies with BSS, leap, parmed topologies using OpenMM ? As the energies are consistent with sander and Adele reports Gromacs works fine with BSS generated topologies the issue may be contained to pmemd.

**Comment by lohedges** [comment]
Just testing this now. I'm also checking what happens if you read/write the `Sire.IO.AmberPrm` topology file with `ParmEd` and use that with `pmemd`, i.e. do the problematic terms get fixed?

**Comment by lohedges** [comment]
If I read/write the `Sire.IO.AmberPrm` topology with `ParmEd`, then the file is identical, other than the version stamp, i.e. the energies would agree when using `pmemd`. There are no warnings etc when parsing the file.

**Comment by lohedges** [comment]
Hmm, `OpenMM` reports different energies for all three topology files, even when performing the minimisation using a Verlet integrator with a _ridiculously_ small time step:

_tLEaP_
```
#"Step" "Time (ps)" "Potential Energy (kJ/mole)" "Kinetic Energy (kJ/mole)" "Total Energy (kJ/mole)" "Temperature (K)"
1 1e-20 -655892.606114115 3263805358977.2705 3263804703084.6646 9454589266.097189 505.85285946764645 0
```

_ParmEd_
```
"Step" "Time (ps)" "Potential Energy (kJ/mole)" "Kinetic Energy (kJ/mole)" "Total Energy (kJ/mole)" "Temperature (K)"
1 1e-20 -682508.7949292829 3336335905771.491 3336335223262.6963 9664695707.432182 402.63291614831616 0
```

_Sire_
```
#"Step" "Time (ps)" "Potential Energy (kJ/mole)" "Kinetic Energy (kJ/mole)" "Total Energy (kJ/mole)" "Temperature (K)" "Box Volume (nm^3)" "Speed (ns/day)" "Time Remaining"
1 1e-20 -682937.7066492349 3408511368084.4966 3408510685146.79 9873773540.270222 402.63291614831616 0 
```

I don't know why `OpenMM` reports a box volume of zero in all cases, or why the temperature agrees for the `Sire` and `ParmEd` topologies, but not for the `tLEaP` one?

**Comment by jmichel80** [comment]
‘If I read/write the Sire.IO.AmberPrm topology with ParmEd, then the file is identical, other than the version stamp, i.e. the energies would agree when using pmemd. There are no warnings etc when parsing the file.’

—> is this a plausible *temporary* bugfix ?

**Comment by lohedges** [comment]
Okay, it looks like `OpenMM` defaults to using the box dimensions from the topology file (which is now deprecated by `AMBER`) and you need to explicitly override this when it is also set in the coordinate file, i.e. 

```python
if inpcrd.boxVectors is not None:
    simulation.context.setPeriodicBoxVectors(*inpcrd.boxVectors)
```

After adding this, then I get the following energies:

_tLEaP_
```
#"Step" "Time (ps)" "Potential Energy (kJ/mole)" "Kinetic Energy (kJ/mole)" "Total Energy (kJ/mole)" "Temperature (K)" "Box Volume (nm^3)"
1 1e-20 -682557.4473130001 3339611920719.1265 3339611238161.679 9674185665.426857 402.63291614831616 0 
```

_ParmEd_
```
#"Step" "Time (ps)" "Potential Energy (kJ/mole)" "Kinetic Energy (kJ/mole)" "Total Energy (kJ/mole)" "Temperature (K)" "Box Volume (nm^3)" 
1 1e-20 -682789.2358909 3328770219475.9766 3328769536686.7407 9642779432.234129 402.63291614831616 0
```

_Sire_
```
#"Step" "Time (ps)" "Potential Energy (kJ/mole)" "Kinetic Energy (kJ/mole)" "Total Energy (kJ/mole)" "Temperature (K)" "Box Volume (nm^3)"
1 1e-20 -682578.6383103809 3322197876923.014 3322197194344.376 9623740674.551035 402.63291614831616 0
```
Again, the energies differ in all cases, although they are now closer.

(I'm also not sure why the temperature and volume records are flipped in the log file. I've checked the state reporter and can't see anything wrong, so perhaps this is something weird within `OpenMM`?)

> —> is this a plausible temporary bugfix ?

No, since the energies are wrong in both cases. Apologies if it wasn't clear, I was saying that `ParmEd` doesn't change the problematic `Sire` topology in any way, whereas it does have differences when reading/writing the original `tLEaP` topology, even though the `pmemd` energies are the same.

**Comment by lohedges** [comment]
Temperature and volume are correct, I was just misreading the number of terms. (The zero is from another reporter that I had truncated from the header.)

**Comment by lohedges** [comment]
Just to note: If we parse `system_parmed.prm7` with `Sire.IO.AmberPrm` and write it back to file, then we generate _exactly_ the same topology as we would do from reading/writing the original `tLEaP` files.

We have this situation...

(Here were are comparing the contents of the files written by the three programs, having read from the format to the left.)

`tLEaP` != `tLEaP` --> `Sire`

`tLEaP` != `tLEaP` --> `ParmEd`

`tLEaP` --> `Sire`  == `tLEaP` --> `ParmEd` --> `Sire`

`tLEaP` --> `Sire` == `tLEaP` --> `Sire` --> `ParmEd`

It's interesting that when parsing the `ParmEd` output (which looks different to our topology and that of `tLEaP`) the file written back to disk _does_ agree with our topology, i.e. the file diff is empty. Similarly, if we use `ParmEd` to parse our topology and write that back to file then the file is unchanged. `ParmEd` is also self-consistent with respect to its own output, i.e. that is unchanged on repeated read/write operations.

**Comment by lohedges** [comment]
Okay, we get energy differences with `pmemd` for a solvated alanine-dipeptide system. (Not sure why I didn't think of testing something simpler before.) At least this will give us something simpler to debug. Just seeing if have a vacuum system anywhere, which would make things even easier.

**Comment by lohedges** [comment]
I just generated one myself using `tLEaP`. In vacuum, the `pmemd` energies agree.

**Comment by lohedges** [comment]
At least, I generated this with `ff14SB` so don't know if the issue is force-field specific.

**Comment by AdeleLip** [comment]
To add to some example systems, I have run 10 ns simulations with `pmemd.cuda` with Sire and tLeap topologies for FXa and BACE. Both systems were unstable when using Sire topologies:
BACE:
![BACE](https://user-images.githubusercontent.com/64790565/115859781-8dd31480-a428-11eb-9ece-fe517445492c.png)
FXa:
![fxa](https://user-images.githubusercontent.com/64790565/115859894-a3483e80-a428-11eb-9467-7163261fcf1a.png)
in both images, the green structure is the starting point, blue is tLeap topology aftern 10 ns and magenta is Sire topology after 10 ns.

**Comment by lohedges** [comment]
I solvated the vacuum alanine-dipeptide in `tLEaP` using a `TIP3PBOX` (following [this](https://ambermd.org/tutorials/basic/tutorial0/index.php#Build_alanine_dipeptide) tutorial) and the resulting structure loaded and saved with `Sire.IO.AmberPrm` gives incorrect energies:

`tLEaP`
```
   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1      -5.6859E+03     1.2214E+01     1.0417E+02     O          23

 BOND    =        0.0540  ANGLE   =        0.3620  DIHED      =        9.6440
 VDWAALS =      711.0740  EEL     =    -6461.0248  HBOND      =        0.0000
 1-4 VDW =        5.0157  1-4 EEL =       48.9355  RESTRAINT  =        0.0000
```

`ParmEd`
```
   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1      -5.6859E+03     1.2214E+01     1.0417E+02     O          23

 BOND    =        0.0540  ANGLE   =        0.3620  DIHED      =        9.6440
 VDWAALS =      711.0740  EEL     =    -6461.0248  HBOND      =        0.0000
 1-4 VDW =        5.0157  1-4 EEL =       48.9355  RESTRAINT  =        0.0000
```

`Sire`
```
   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1      -5.7492E+03     1.2213E+01     1.0417E+02     O          23

 BOND    =        0.0540  ANGLE   =        0.3620  DIHED      =        9.6440
 VDWAALS =      711.0740  EEL     =    -6461.0248  HBOND      =        0.0000
 1-4 VDW =        3.3048  1-4 EEL =      -12.6096  RESTRAINT  =        0.0000
```

I feel like we're getting closer, although I'm not sure what could be different in the treatment of water molecules between `sander` and `pmemd`. Note that I am doing all this directly with Sire, so it's not a result of BioSimSpace swapping the water topology behind the scenes. (We don't do this if it already matches the AMBER template anyway.)

**Comment by jmichel80** [comment]
@lohedges the fact the the 1,4 terms are always less positive in Sire suggests that a number of 1,4 interactions are skipped or zeroed due to an incorrect scaling factor. 
You could hack parm/parm10.dat and parm/frcmod.ff14SB to remove all multidihedral terms (all the dihedral entries with a negative number in the 5th column). Then setup a new topology with tleap and see if that makes the 1,4 energies agree between codes. If the issue still persists then it must be an indexing error, though I don't see why at first it would happen after addition of TIP3P water (since those molecules don't have 1,4 terms).

**Comment by lohedges** [comment]
I've just checked Sire 2019.1.0 for my own sanity and the issue is still present. It seems like it is something that has been present with the parser from the outset, rather than anything to do with recent fixes for incorrect POINTERS, etc.

I"ll try your suggestions and report back if I make any progress.

**Comment by lohedges** [comment]
I've removed multidihedral terms as suggested and it makes things worse. The `1-4 VDW` is similar, but the `1-4 EEL` term is even less positive than before. Both the original `tLEaP` and `ParmEd` topologies still give the same energies. (Different to before, but the same as each other.)

**Comment by jmichel80** [comment]
I have run tests on alanine dipeptide in vacuum with the following inputs 
[vacuum.zip](https://github.com/michellab/Sire/files/6365720/vacuum.zip)

and I observe inconsistencies in single point energies in vacuum with pmemd. 

(to reproduce extract the attached zip, edit COMPARE_SP to point to the paths for sander/pmemd and execute the bash script. 

```
================= SANDER BSS single point vacuum ==========


   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1      -1.3334E+01     5.4688E+00     1.8394E+01     N           7

 BOND    =        0.0206  ANGLE   =        0.3620  DIHED      =        9.6440
 VDWAALS =        2.8120  EEL     =      -80.1238  HBOND      =        0.0000
 1-4 VDW =        5.0157  1-4 EEL =       48.9355  RESTRAINT  =        0.0000
================= SANDER leap single point vacuum ==========


   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1      -1.3334E+01     5.4688E+00     1.8394E+01     N           7

 BOND    =        0.0206  ANGLE   =        0.3620  DIHED      =        9.6440
 VDWAALS =        2.8120  EEL     =      -80.1238  HBOND      =        0.0000
 1-4 VDW =        5.0157  1-4 EEL =       48.9355  RESTRAINT  =        0.0000
================= PMEMD BSS single point vacuum ==========


   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1      -6.0261E+01     1.3445E+01     4.1055E+01     HH32       21

 BOND    =        0.0206  ANGLE   =        0.3620  DIHED      =        9.6440
 VDWAALS =        2.8120  EEL     =      -63.7951  HBOND      =        0.0000
 1-4 VDW =        3.3048  1-4 EEL =      -12.6096  RESTRAINT  =        0.0000
================= PMEMD leap single point vacuum ==========


   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1       2.9947E+00     1.3750E+01     4.3485E+01     HH32       21

 BOND    =        0.0206  ANGLE   =        0.3620  DIHED      =        9.6440
 VDWAALS =        2.8120  EEL     =      -63.7951  HBOND      =        0.0000
 1-4 VDW =        5.0157  1-4 EEL =       48.9355  RESTRAINT  =        0.0000
```

Note that pmemd and sander EEL energies differ because PMEMD seems to still use PME in a non periodic input whereas sander uses coulombic without cutoff.

**Comment by jmichel80** [comment]
An update - I deleted the entries %FLAG SCEE_SCALE_FACTOR and %FLAG SCNB_SCALE_FACTOR from both leap an bss topologies and repeated the single point energy calculations. I obtained identical results, with both codes indicating they used default SCEE/SCNB values. 
Therefore the discrepancy in the calculation of 1,4 non bonded terms is not due to the 1,4 scaling factors.

**Comment by lohedges** [comment]
Hmm, interesting. I wonder what was different with our vacuum systems. Perhaps I had simply coped across the wrong file to the cluster in my debugging haze.

**Comment by jmichel80** [comment]
Another update (hold your breath). 
If you delete all the FLAG entries in system_bss.prm7 or system_leap.parm7 below FLAG JOIN ARRAY pmemd gives the same single point energy components as with sander ! 

```
================= SANDER BSS single point vacuum ==========


   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1       2.9947E+00     1.3750E+01     4.3485E+01     HH32       21

 BOND    =        0.0206  ANGLE   =        0.3620  DIHED      =        9.6440
 VDWAALS =        2.8120  EEL     =      -63.7951  HBOND      =        0.0000
 1-4 VDW =        5.0157  1-4 EEL =       48.9355  RESTRAINT  =        0.0000
================= SANDER leap single point vacuum ==========


   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1       2.9947E+00     1.3750E+01     4.3485E+01     HH32       21

 BOND    =        0.0206  ANGLE   =        0.3620  DIHED      =        9.6440
 VDWAALS =        2.8120  EEL     =      -63.7951  HBOND      =        0.0000
 1-4 VDW =        5.0157  1-4 EEL =       48.9355  RESTRAINT  =        0.0000
================= PMEMD BSS single point vacuum ==========


   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1       2.9947E+00     1.3750E+01     4.3485E+01     HH32       21

 BOND    =        0.0206  ANGLE   =        0.3620  DIHED      =        9.6440
 VDWAALS =        2.8120  EEL     =      -63.7951  HBOND      =        0.0000
 1-4 VDW =        5.0157  1-4 EEL =       48.9355  RESTRAINT  =        0.0000
================= PMEMD leap single point vacuum ==========


   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1       2.9947E+00     1.3750E+01     4.3485E+01     HH32       21

 BOND    =        0.0206  ANGLE   =        0.3620  DIHED      =        9.6440
 VDWAALS =        2.8120  EEL     =      -63.7951  HBOND      =        0.0000
 1-4 VDW =        5.0157  1-4 EEL =       48.9355  RESTRAINT  =        0.0000
```

**Comment by jmichel80** [comment]
Ok strike that,  parsing error with my script...

**Comment by jmichel80** [comment]
Another update. I zeroed the charges on all atoms but the first 1,4 pair (ACE1:H1 and ACE1:O). This gives the same 1,4 EEL single point energy between sander and pmemd with a BSS topology. 

I also removed all negative indices in the 3rd entry of each dihedrals in both topologies. This would cause over-counting of 1,4 interactions with multi term dihedrals.  This manifests as a change in 1,4 pmemd energies for both topologies, but NOT for sander. 

```
================= SANDER BSS single point vacuum ==========


   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1      -1.3334E+01     5.4688E+00     1.8394E+01     N           7

 BOND    =        0.0206  ANGLE   =        0.3620  DIHED      =        9.6440
 VDWAALS =        2.8120  EEL     =      -80.1238  HBOND      =        0.0000
 1-4 VDW =        5.0157  1-4 EEL =       48.9355  RESTRAINT  =        0.0000
================= SANDER leap single point vacuum ==========


   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1      -1.3334E+01     5.4688E+00     1.8394E+01     N           7

 BOND    =        0.0206  ANGLE   =        0.3620  DIHED      =        9.6440
 VDWAALS =        2.8120  EEL     =      -80.1238  HBOND      =        0.0000
 1-4 VDW =        5.0157  1-4 EEL =       48.9355  RESTRAINT  =        0.0000
================= PMEMD BSS single point vacuum ==========


   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1      -6.0532E+01     1.4265E+01     4.1055E+01     HH32       21

 BOND    =        0.0206  ANGLE   =        0.3620  DIHED      =        9.6440
 VDWAALS =        2.8120  EEL     =      -63.7951  HBOND      =        0.0000
 1-4 VDW =        5.6818  1-4 EEL =      -15.2570  RESTRAINT  =        0.0000
================= PMEMD leap single point vacuum ==========


   NSTEP       ENERGY          RMS            GMAX         NAME    NUMBER
      1       4.4690E+01     1.5820E+01     4.3485E+01     HH32       21

 BOND    =        0.0206  ANGLE   =        0.3620  DIHED      =        9.6440
 VDWAALS =        2.8120  EEL     =      -63.7951  HBOND      =        0.0000
 1-4 VDW =        8.5638  1-4 EEL =       87.0830  RESTRAINT  =        0.0000  
```

What this indicates is that sander (at least the version in AmberTools20) uses its own nternal logic to work out which 1,4 pairs to consider for non bonded calculations,  whereas pmemd relies on a particular layout of FLAG DIHEDRALS_INC_HYDROGEN and FLAG DIHEDRALS_WITHOUT_HYDROGEN to do so....

**Comment by lohedges** [comment]
Thanks, @jmichel80 this is really helpful. I feel like we're close to getting to the bottom of it. I'll check to see how those records are generated in Sire and also see how ParmEd writes them in case there are any clues as to the specific ordering that is required.

**Comment by lohedges** [comment]
Just to say that I re-tried removing the dihedral sorting and I still get the same incorrect energies with `pmemd`. (Before I got different incorrect energies, but I disabled sorting for all terms, rather than just the dihedrals.)

**Comment by lohedges** [comment]
This is the [code](https://github.com/michellab/Sire/blob/devel/corelib/src/libs/SireIO/amberprm.cpp#L3056) that supposedly takes care of the double counting issue:

```cpp
        auto ignore_14 = [](QVector<qint64> &dihs)
        {
            QSet<QPair<qint64, qint64> > seen_dihedrals;
            seen_dihedrals.reserve(dihs.count()/5);

            for (int i=0; i<dihs.count(); i+=5)
            {
                //NB14 must be computed only the first dihedral found that has this pair,
                //otherwise we would risk Amber double-counting this parameter
                const QPair<qint64, qint64> nb14pair(dihs[i], dihs[i+3]);

                if (seen_dihedrals.contains(nb14pair))
                {
                    dihs[i+2] = -dihs[i+2];
                }
                else if (dihs[i+2] > 0)
                {
                    seen_dihedrals.insert(nb14pair);
                }
            }
        };

        ignore_14(all_dihs_inc_h);
        ignore_14(all_dihs_exc_h);

```

Disabling this (by commenting out the last two lines) makes no difference to the `sander` results, but changes the `1-4 VDW` and `1-4 EEL` values. (Still incorrect, but different.) 

I am assuming that `pmemd` requires the NB14 to be set for a specific dihedral to be set. The source code suggests the _last_ entry if multiple dihedrals contain the same pair, but inverting the order of the loop makes no difference to the result. I then wondered if it wouldn't work if we sorted dihedrals, but that makes no difference either. I've also tried disabling parallelisation while parsing (in case that changed the order), but again no difference.

I'm still getting my head around the logic of the code above, e.g. why the 3rd atom index in the dihedral is inverted if ignored. Perhaps there is something simple that is wrong here.

I was also wondering if, following the above, we then need to ourder the `DIHEDRALS_INC_HYDROGEN` and 
`DIHEDRALS_WITHOUT_HYDROGEN` such that the terms that aren't ignored are last? I'll check the ordering of the original `tLEaP` files for your vacuum system to see if it throws up any clues.

I feel like we're close.

**Comment by lohedges** [comment]
Comparing the two files, it looks like all of the `tLEaP` entries _are_ sorted, whereas ours aren't, even though sorting is turned on. Perhaps something funny is going on with the sorting, I'll try to reproduce the same ordering.

**Comment by lohedges** [comment]
Scratch that, looking at the files the wrong way round :man_facepalming:

**Comment by jmichel80** [comment]
I also compared the 1,4 pairs list implied by the dihedral entries and found exactly the same pairs albeit in a different order. Everything suggests the issue is the different ordering. Perhaps causing incorrect lookup of charges (DIH is consistent so the coordinates are consistent). We may be able to work this out by recompiling pmemd with verbose/debugging statements.

**Comment by lohedges** [comment]
Looking at the files, the obvious difference is that the impropers always come after dihedrals, whereas the Sire output is interleaved. It might be a red herring, but at least it's something easy to test.

**Comment by lohedges** [comment]
Red herring it is.

**Comment by jmichel80** [comment]
I found the problem by digging in pmemd's source code. 
The problem is that the bss topology of the alanine dipeptide the 3rd entry of the DIHEDRAL_PERIODICITY section is 0

> %FLAG DIHEDRAL_PERIODICITY
> %FORMAT(5E16.8)
>   2.00000000E+00  2.00000000E+00  0.00000000E+00  3.00000000E+00  3.00000000E+00
>   1.00000000E+00  2.00000000E+00  2.00000000E+00  3.00000000E+00  3.00000000E+00
>   1.00000000E+00  3.00000000E+00  1.00000000E+00  3.00000000E+00  2.00000000E+00
>   2.00000000E+00  1.00000000E+00  2.00000000E+00

I don't know why we could have torsions with 0 periodicity. 

This causes pmemd to disable this non-bonded pair in subroutine calc_dihedral_parms. 

Simply  changing this to entry to a positive number gives agreement of single point energies between the bss and leap topologies. 

The torsional energies would still be consistent because the dihedral force constant is 0 for this entry. 

@lohedges now just need to figure out why BSS writes down null dihedral parameters..

**Comment by lohedges** [comment]
Fantastic! I'll take a look when I fire the computer up.

**Comment by lohedges** [comment]
I've compared the dihedral and improper terms from the `Sire.MM.AmberParams` object contructed on read to the individual `Sire.MM.AmberDihedral` objects that are reconstructed using `Sire.CAS` when we convert back to AMBER format from a `Sire.System`. I don't see any zero periodicities, other null dihedrals.

The issue you found from digging through the source code does seem to be the same issue that was flagged by the `ParmEd` team, which I posted above. I'll copy here for reference:

```python
# pmemd likes to skip torsions with periodicities of 0, which may be
        # present as a way to hack entries into the 1-4 pairlist. See
        # https://github.com/ParmEd/ParmEd/pull/145 for discussion. The solution
        # here is to simply set that periodicity to 1.
        inst._cleanup_dihedrals_with_periodicity_zero()
...

def _cleanup_dihedrals_with_periodicity_zero(self):
        """
        For torsions with only a single term and a periodicity set to 0, make sure pmemd still
        properly recognizes the necessary exception parameters. update_dihedral_exclusions will
        make sure that if a dihedral has a type pn0 *and* ignore_end is set to False (which means
        that it is required to specify exclusions), then it is the *only* torsion between those
        atoms in the system. This allows us to scan through our dihedrals, look for significant
        terms that have pn==0, and simply add another dihedral with pn=1 and k=0 to ensure that
        pmemd will always get that exception correct
        """
        new_dihedrals = []
        for dih in self.dihedrals:
            if dih.ignore_end or dih.type.per != 0:
                continue
            # If we got here, ignore_end must be False and out periodicity must be 0. So add
            # another dihedral
            dt = DihedralType(0, 1, 0, dih.type.scee, dih.type.scnb, list=self.dihedral_types)
            self.dihedral_types.append(dt)
            new_dihedrals.append(
                Dihedral(dih.atom1, dih.atom2, dih.atom3, dih.atom4, improper=dih.improper,
                         ignore_end=False, type=dt)
            )
            # Now that we added the above dihedral, we can start ignoring the end-group interactions
            # on this dihedral
            dih.ignore_end = True
        if new_dihedrals:
            self.dihedrals.extend(new_dihedrals)
```

I'll try to figure out where the zero periodicity is being set in our code. I'll also try setting the periodicity of null terms to one to see if that fixes things.

**Comment by lohedges** [comment]
Okay, I've fixed it. The solution was to set a default periodicity of one for the `Sire.MM.AmberDihPart` constructor [here](https://github.com/michellab/Sire/blob/devel/corelib/src/libs/SireMM/amberparams.h#L227). Although I had tested for the `ParmEd` issue previously, the `Sire.MM.AmberDihedral` constructor wasn't even getting to the part where I was performing the check, since there were _no_ terms, rather than a single terms with zero periodicity. (By default we construct a `Sire.MM.AmberDihedral` with a single `Sire.MM.AmberDihPart` term with zero values for force constant, periodicity, and phase.)

The energies now agree for all systems that I've tested with `pmemd`. I'll just check our unit test suite to see if this has any knock-on affects, then push the update.

**Comment by lohedges** [comment]
This was a single character fix, i.e. 0 --> 1. Possibly the most tortuous debugging I have ever done, followed by the easiest fix.

**Comment by jmichel80** [comment]
A war story for your memoir !

**Comment by jmichel80** [comment]
@AdeleLip could you repeat your test MD sims on PTP1B and BACE with the new code once available?

**Comment by AdeleLip** [comment]
Yes, I'll do that now and update

**Comment by chryswoods** [comment]
Well done @lohedges  - that is the most tortuous debugging with the easiest single character fix that I think I've ever seen. Seriously good work! :-)

**Comment by AdeleLip** [comment]
both systems were stable on the timescale of 5 ns! I will do a full set up from PDB file for PTP1B and a longer MD run to confirm for sure, but it looks solved.

**Comment by lohedges** [comment]
Fantastic. thanks for the update. I think we all need a holiday after this!
